User should be able to set the email address

sharidas opened the pull request:
User logged in should be able to set his/her
email address.

Signed-off-by: Sujith H <sharidasan@owncloud.com>

<!--
Thanks for submitting a change to ownCloud!

This is the bug tracker for the Server component. Find other components at https://github.com/owncloud/core/blob/master/.github/CONTRIBUTING.md#guidelines

For fixing potential security issues please see https://owncloud.org/security/

To make it possible for us to get your change reviewed and merged please fill out below information carefully.

Please note that any kind of change first has to be submitted to the master branch which holds the next major version of ownCloud.

We will carefully discuss if your change can or has to be backported to stable branches.
-->

## Description
<!--- Describe your changes in detail -->
Users were not able to set the email address due to the regression caused at the commit: https://github.com/owncloud/core/commit/b73b488cc91803d2236a9f1ff7ac7d78b6b1b1fa

## Related Issue
<!--- This project only accepts pull requests related to open issues -->
<!--- If suggesting a new feature or change, please discuss it in an issue first -->
<!--- If fixing a bug, there should be an issue describing it with steps to reproduce -->
<!--- Please link to the issue here: -->
https://github.com/owncloud/core/issues/27667

## Motivation and Context
<!--- Why is this change required? What problem does it solve? -->
Added an extra condition to check if the user logged in is trying to set the email. And it should be possible for the user to set the email address.

## How Has This Been Tested?
<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->
- [x] Created user `user1`
- [x] Logged in as `user1`
- [x] And then set the email address
- [x] Verified the confirmaition email has arrived
- [x] Use the confirmation link got in the email to verify the email address.
- [x] Refresh the `user1` settings page to see if the email is shown in the email text box.
- [x] Verified that email is updated.

## Screenshots (if appropriate):

## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the **CONTRIBUTING** document.
- [x] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

sharidas committed:
User should be able to set the email address

User should be able to set his/her
email address after the email address is validated.

Signed-off-by: Sujith H <sharidasan@owncloud.com>

SergioBertolinSG commented:
It works when validating the email. 

@PVince81 should this work also when the email is not validated? If this is a case I think there should be a message telling the user that the email will be removed if it is not validated inmediately.

PVince81 commented on settings/Controller/UsersController.php:
```diff
@@ -792,18 +792,21 @@ public function sendEmail($userId, $mailAddress) {
 			($this->groupManager->getSubAdmin()->isSubAdmin($this->userSession->getUser()) &&
-				$this->groupManager->getSubAdmin()->isUserAccessible($this->userSession->getUser(), $user))) {
+				$this->groupManager->getSubAdmin()->isUserAccessible($this->userSession->getUser(), $user)) ||
+				($this->userSession->getUser()->getUID() === $id)) {
```
note that the current user does not need to be logged in to validate their email. they just need to open the link.

will it work ?

sharidas replied:
Yes, So if the user tries to open the link, after clicking the link, this will work.

SergioBertolinSG replied:
It does not work without login. If there is no session active it opens the login page, if there is another user logged in, it fails. With a logged in user it works correctly.

PVince81 replied:
Can you try with the old version before the regression ? I don't remember whether we decided that login is required or not

sharidas replied:
@PVince81  In the 10.0.0 version where this was working before, login is required.

PVince81 replied:
Ok, then let's keep that as the expected behavior

sharidas replied:
@PVince81 Ok, then I will update the PR again. Because I removed lines https://github.com/owncloud/core/blob/master/settings/Controller/UsersController.php#L827-L832. Will bring back the unit test for the same which was removed.

PVince81 replied:
Thanks. In general, when in doubt: when a regression occurs, make it work like it worked before the regression.

sharidas commented:
Updated further. Removed the check which was done to verify if the user is logged in when the link from email is clicked. And hence removed the unit test which was associated with that test. Let me know if this looks ok. Also I have modified the string message.

sharidas commented:
Keeping the old behaviour as such, updated the PR.

SergioBertolinSG commented:
@sharidas testing this last version of this PR and accepting that the user must be logged to verify his address, I've seen something:

It works when you open the link in the same browser were you have the session open. But opening the link in a different browser without a session, fails. It opens the login page, and even after login with the user changing its email, it fails.

@PVince81 is this expected?

PVince81 commented:
> even after login with the user changing its email, it fails.

this is a bug. Maybe the redirect doesn't properly redirect to the mail link.

sharidas commented:
@SergioBertolinSG Here is how I tested:
- [x] Open firefox and then try to login as `user1`
- [x] set the email address for `user1`
- [x] Now try to open the link obtained from the email ( sent to the respective email, in my case I used gmail ) in another browser. I used Chrome.
- [x] It opens the login page. And I am able to get the email address set for `user1`

For me it works.

sharidas commented:
![gifrecord_2018-01-30_161528](https://user-images.githubusercontent.com/3600427/35562535-afbf9938-05d9-11e8-837e-947d2a8b3bfe.gif)

Hope this helps.

SergioBertolinSG commented:
Retested, it was a problem with the email site I was using which was cutting the link. it works 👍 .

SergioBertolinSG reviewed (approved):
Just tested it and it works. No code review.

PVince81 commented on settings/Controller/UsersController.php:
```diff
@@ -620,7 +620,7 @@ public function setMailAddress($id, $mailAddress) {
 							'username' => $id,
 							'mailAddress' => $mailAddress,
-							'message' => (string) $this->l10n->t('An email has been sent to this address for confirmation')
+							'message' => (string) $this->l10n->t('An email has been sent to this address for confirmation. Until email is verified this address will not be set.')
```
@phil-davis quick wording check ?

phil-davis replied:
Could just add a "the":
```
Until the email is verified this address will not be set.
```

sharidas replied:
@phil-davis  Thanks for the feedback. Updated. Yah that's true for CI

PVince81 reviewed (approved):
:+1: code looks fine.

If @phil-davis has no objection to the wording this can be merged.

phil-davis reviewed (approved):
Good to merge when CI passes

phil-davis commented:
Backport stable10 #30319
